packages/samples/src/sample-snapshot-testing.ts

```diff
@@ -0,0 +1,167 @@
+import {
+  CompilerHost,
+  NodeHost,
+  ResolveCompilerOptionsOptions,
+  compile,
+  getDirectoryPath,
+  joinPaths,
+  resolveCompilerOptions,
+  resolvePath,
+} from "@typespec/compiler";
+import { expectDiagnosticEmpty } from "@typespec/compiler/testing";
+import { fail, ok, strictEqual } from "assert";
+import { readdirSync } from "fs";
+import { mkdir, readFile, readdir, writeFile } from "fs/promises";
+
+const shouldUpdateSnapshots = process.env.RECORD === "true";
+
+export interface SampleSnapshotTestOptions {
+  /**  Sample root directory. */
+  sampleDir: string;
+
+  /** Output directory for snapshots. */
+  outputDir: string;
+
+  /** Folders to exclude from testing. */
+  exclude?: string[];
+
+  /** Override the emitters to use. */
+  emit?: string[];
+}
+
+export function defineSampleSnaphotTests(config: SampleSnapshotTestOptions) {
+  resolveSamples(config).forEach((samples) => defineSampleSnaphotTest(config, samples));
+}
+
+function defineSampleSnaphotTest(config: SampleSnapshotTestOptions, sample: Sample) {
+  it(sample.name, async () => {
+    const host = createSampleSnapshotTestHost(config);
+
+    const overrides: Partial<ResolveCompilerOptionsOptions["overrides"]> = {
+      outputDir: "/out",
+    };
+    if (config.emit) {
+      overrides.emit = config.emit;
+    }
+    const [options, diagnostics] = await resolveCompilerOptions(host, {
+      entrypoint: sample.fullPath,
+      overrides,
+    });
+    expectDiagnosticEmpty(diagnostics);
+
+    const emit = options.emit;
+    if (emit === undefined || emit.length === 0) {
+      fail(
+        `No emitters configured for sample "${sample.name}". Make sure the  config at: "${options.config}" is correct.`
+      );
+    }
+
+    const program = await compile(host, sample.fullPath, options);
+    expectDiagnosticEmpty(program.diagnostics);
+
+    const outputDir = resolvePath(config.outputDir, sample.name);
+
+    if (shouldUpdateSnapshots) {
+      try {
+        await host.rm(outputDir, { recursive: true });
+      } catch (e) {}
+      await mkdir(outputDir, { recursive: true });
+
+      for (const [filename, content] of host.outputs.entries()) {
+        const snapshotPath = resolvePath(outputDir, filename);
+
+        try {
+          await mkdir(getDirectoryPath(snapshotPath), { recursive: true });
+          await writeFile(snapshotPath, content);
+        } catch (e) {
+          throw new Error(`Failure to write snapshot: "${snapshotPath}"\n Error: ${e}`);
+        }
+      }
+    } else {
+      for (const [filename, content] of host.outputs.entries()) {
+        const snapshotPath = resolvePath(outputDir, filename);
+        let existingContent;
+        try {
+          existingContent = await readFile(snapshotPath);
+        } catch (e: unknown) {
+          if (typeof e === "object" && e !== null && "code" in e && e.code === "ENOENT") {
```

Comment:
With this change, we should change the ci not to run regen-samples anymore, but run sample tests as part of tests, right?

Reply:
We still want to regen the sample in the CI to make sure the `CLEAN_SAMPLES` doesn't delete a sample that wasn't delete in the PR.

I am actually not sure we want to run them as "test" in the CI. Just regen the sample I think is sufficient .

Reply:
ok so I updated the snapshot runner to automatically detect all the files that were not generated during a full test run so it would complain if you leave ghost files. As long as you don't reorganize the output folders where then you might have to delete manually but that would probably already been the case before.